src/language-markdown/printer-markdown.js

```diff
@@ -526,21 +561,244 @@ function getAncestorNode(path, typeOrTypes) {
+ * you would replace `"\n"` with `" "`. (`"You might want to use Prettier."`)
+ *
+ * However, you should note that Chinese and Japanese does not use U+0020 Space to divide words, so U+000A End of Line must not be replaced with it.
+ * Behavior in other languages (e.g. Thai) will not be changed because there are too much things to consider. (PR welcome)
+ *
+ * @param {AstPath} path path of given node
+ * @param {WhitespaceValue} value value of given node (typically `" "` or `"\n"`)
+ * @param {AdjacentNodes | undefined} adjacentNodes adjacent sibling nodes of given node
+ * @returns {boolean} `true` if given node can be converted to space, `false` if not (i.e. newline or empty character)
+ */
+function canBeConvertedToSpace(path, value, adjacentNodes) {
+  // "\n" or " ", of course " " always can be converted to Space
+  if (value !== "\n") {
+    return true;
+  }
+  // no adjacent nodes
+  if (!adjacentNodes) {
+    return true;
+  }
+  // e.g. " \nletter"
+  if (!adjacentNodes.previous || !adjacentNodes.next) {
+    return true;
+  }
+  const previousKind = adjacentNodes.previous.kind;
+  const nextKind = adjacentNodes.next.kind;
+  // "\n" between not western or Korean (han, kana, CJK punctuations) characters always can converted to Space
+  // Korean hangul simulates latin words; see #6516 (https://github.com/prettier/prettier/issues/6516)
+  if (
+    isWesternOrKoreanLetter(previousKind) &&
+    isWesternOrKoreanLetter(nextKind)
+  ) {
+    return true;
+  }
+  // han & hangul: same way preferred
+  if (
+    (previousKind === KIND_K_LETTER && nextKind === KIND_CJ_LETTER) ||
+    (nextKind === KIND_K_LETTER && previousKind === KIND_CJ_LETTER)
+  ) {
+    return true;
+  }
+  // Do not convert it to Space when:
+  if (
+    // Shall not be converted to Space around CJK punctuation
+    previousKind === KIND_CJK_PUNCTUATION ||
+    nextKind === KIND_CJK_PUNCTUATION ||
+    // "\n" between CJ always SHALL NOT be converted to space
+    // "\n" between Korean and CJ is better not to be converted to space
+    (isCJK(previousKind) && isCJK(nextKind))
+  ) {
+    return false;
+  }
+  const previousLastChar = adjacentNodes.previous.value?.at(-1);
+  const nextFirstChar = adjacentNodes.next.value?.[0];
+
+  // From here down, only line breaks between CJ and alphanumeric characters are covered.
+
+  // Convert newline between CJ and specific symbol characters (e.g. ASCII punctuation)  to Space.
+  // e.g. :::\n句子句子句子\n::: → ::: 句子句子句子 :::
+  //
+  // Note: Line breaks like "(\n句子句子\n)" or "句子\n." by Prettier are suppressed in `isBreakable(...)`.
+  if (
+    lineBreakBetweenTheseAndCJKConvertToSpaceSymbolSet.has(nextFirstChar) ||
+    lineBreakBetweenTheseAndCJKConvertToSpaceSymbolSet.has(previousLastChar)
+  ) {
+    return true;
+  }
+  // Converting newline between CJ and non-ASCII punctuation to Space does not seem to be better in many cases. (PR welcome)
```

Comment:
Could you add an example or two here of such situations where it doesn't seem better?

Reply:
@thorn0 

```js
  // e.g.
  // 1. “ア〜エの中から1つ選べ。”
  // "〜" (U+301C) belongs to Pd, and "\n" in "ア〜\nエの中から1つ選べ。" must not be converted to Space.
  // 2. “これはひどい……なんと汚いコミットログなんだ……”
  // "…" (U+2026) belongs to Po, and "\n" in "これはひどい……\nなんと汚いコミットログなんだ……" must not, either.
```